src/unix/process.c

```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
 }
 #endif
 
+#if defined(__APPLE__)
+
+typedef struct uv__posix_spawn_fncs_tag {
+  struct {
+    int (*set_uid_np)(const posix_spawnattr_t *, uid_t);
+    int (*set_gid_np)(const posix_spawnattr_t *, gid_t);
+    int (*set_groups_np)(const posix_spawnattr_t*, int, gid_t*, uid_t);
+  } spawnattr;
+
+  struct {
+    int (*addchdir_np)(const posix_spawn_file_actions_t *, const char *);
+  } file_actions;
+} uv__posix_spawn_fncs_t;
+
+int uv__spawn_use_posix_spawn(uv__posix_spawn_fncs_t* fncs) {
+  /* Try to locate all non-portable functions at runtime */
+  fncs->spawnattr.set_uid_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_uid_np");
+  fncs->spawnattr.set_gid_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_gid_np");
+  fncs->spawnattr.set_groups_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_groups_np");
+  fncs->file_actions.addchdir_np = dlsym(RTLD_DEFAULT, "posix_spawn_file_actions_addchdir_np");
+
+  /* If any of these is missing, just don't use posix_spawn */
```

Comment:
These are all fairly new (10.15), which means we currently can't test this code on libuv CI (or use it on other platforms). Could we instead selectively enable the posix_spawn code for the common cases were we don't need to use any of these? (return ENOSYS from uv__spawn_and_init_child_posix_spawn, then have that bubble up to uv__spawn_and_init_child where it can use the fork-exec next strategy)

Reply:
I've implemented this alternative.